kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt

```diff
@@ -59,46 +72,67 @@ internal class WorkflowRunnerViewModel<OutputT : Any>(
     }
   }
 
-  private val subs = CompositeDisposable()
+  private val snapshotJob = GlobalScope.launch(context) {
+    workflowHost.renderingsAndSnapshots
+        .map { it.snapshot }
+        .collect { lastSnapshot = it }
+  }
 
-  @Suppress("EXPERIMENTAL_API_USAGE")
-  private val updates =
-    workflowUpdates.updates.asObservable(Dispatchers.Unconfined)
-        .doOnNext { lastSnapshot = it.snapshot }
-        .replay(1)
-        .autoConnect(1) { subs.add(it) }
+  private var workflowJob: Job? = null
 
   private var lastSnapshot: Snapshot = Snapshot.EMPTY
 
-  override val renderings: Observable<out Any> = updates.map { it.rendering }
+  @UseExperimental(ExperimentalCoroutinesApi::class)
+  override val renderings: Observable<out Any> = workflowHost.renderingsAndSnapshots
+      // Lazily start the workflow the first time someone starts listening. Like autoConnect(1).
+      .onCollect {
+        // This call is idempotent and will always return the same Job, so we can just call it
+        // every time.
+        workflowJob = workflowHost.start()
+      }
+      .map { it.rendering }
+      .asObservable()
 
-  override val output: Flowable<out OutputT> =
-    // Buffer on backpressure so outputs don't get lost.
-    updates.toFlowable(BUFFER)
-        .filter { it.output != null }
-        .map { it.output!! }
-        // DON'T replay, outputs are events.
-        .publish()
-        // Subscribe upstream immediately so we immediately start getting notified about outputs.
-        // If [renderings] triggers the upstream subscription before we do, if the workflow emits
-        // an output immediately we might not see it.
-        .autoConnect(0) { subs.add(it) }
+  @UseExperimental(ExperimentalCoroutinesApi::class)
+  override val output: Flowable<out OutputT> = workflowHost.outputs
+      // Lazily start the workflow the first time someone starts listening. Like autoConnect(1).
+      .onCollect {
+        // This call is idempotent and will always return the same Job, so we can just call it
+        // every time.
+        workflowJob = workflowHost.start()
+      }
+      .asFlowable()
 
   override fun onCleared() {
-    // Has the side effect of closing the updates channel, which in turn
-    // will fire any tear downs registered by the root workflow.
-    subs.clear()
+    val cancellationException = CancellationException("WorkflowRunnerViewModel cleared.")
+    snapshotJob.cancel(cancellationException)
+    workflowJob?.cancel(cancellationException)
   }
 
   override fun onSaveInstanceState(outState: Bundle) {
     outState.putParcelable(BUNDLE_KEY, PickledWorkflow(lastSnapshot))
   }
 
+  @TestOnly
+  internal fun clearForTest() = onCleared()
+
+  @TestOnly
+  internal fun getLastSnapshotForTest() = lastSnapshot
+
   private companion object {
     val BUNDLE_KEY = WorkflowRunner::class.jvmName + "-workflow"
   }
 }
+
+/**
+ * Invokes `block` every time a new collector begins collecting this [Flow].
```

Comment:
Whoa, I didn't realize you had built this method yourself. Neat. Seems like it should be built in: worth submitting?

Reply:
Yes, I plan to.